### PR TITLE
1366 1337 - Add covid contact and covid infection LOCF

### PR DESCRIPTION
**Pull request by slamothe99**
## Pre-review checklist

I have ensured that:

- [ ] Code runs successfully in the development environment.
- [ ] New code is tested to prove that it works. If not, justify reasoning.
- [ ] New code is documented using [type hinting](https://realpython.com/lessons/type-hinting/) and following the [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) docstring format.
- [ ] New code follows the project naming standards.

## Post-review checklist

As I merge these changes, I have:

- [ ] Updated the related Jira ticket, including approximate time spent (including review).
- [ ] Moved the related Jira ticket to Done.

### CODE REVIEW

File: cishouseholds/impute.py [hunk 1]
```diff
@@ -10,6 +10,63 @@
...
+    transformed_df = transformed_df.withColumn(
+        f"{date_column}_ref",
+        F.when(F.col(f"{date_column}_ref") <= F.col(visit_date_column), F.col(f"{date_column}_ref")).otherwise(None),
+    )
+
+    window = Window.partitionBy(participant_id_column, visit_date_column, f"{date_column}_ref")
+    transformed_df = update_column_values_from_map(transformed_df, type_column, hierarchy_map)
+    transformed_df = transformed_df.withColumn(type_column, F.min(type_column).over(window))
+
+    reverse_map = {}
```
Comment (foster999) [inline cishouseholds/impute.py:50]: It's more pythonic and concise to reverse a dict using `{value: key for key, value in hierarchy_map.items()}`

File: cishouseholds/impute.py [hunk 2]
```diff
@@ -10,6 +10,63 @@
...
+    df: DataFrame,
+    participant_id_column: str,
+    date_column: str,
+    visit_date_column: str,
+    type_column: str,
+    contact_column: str,
+    hierarchy_map: dict,
+):
+    """
+    impute think had covid and associated variables
```
Comment (foster999) [inline cishouseholds/impute.py:26]: Please add a description of how the hierarchy_map should be used

File: cishouseholds/impute.py [hunk 3]
```diff
@@ -10,6 +10,63 @@
...
+    Parameters
```
Comment (foster999) [inline cishouseholds/impute.py:27]: Don't worry about adding these to the docstring unless you need to provide additional description of each parameters, as the names can be seen in the `def` above

File: tests/impute/test_impute_think_had_covid.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
+import pyspark.sql.functions as F
+from chispa import assert_df_equality
+
+from cishouseholds.impute import impute_think_had_covid
+
+
+def test_impute_think_had_covid(spark_session):
+    input_df = spark_session.createDataFrame(
+        data=[
+            (1, "2020-03-12", "2020-12-20", "A", "Yes"),
```
Comment (foster999) [inline tests/impute/test_impute_think_had_covid.py:10]: Please add some notes to say what is being checked for each participant here

**Review by foster999 — CHANGES_REQUESTED** [review]
Some minor changes suggested here in comments, but also see the minimal test case that I've added.

*_type in the second row should take the value from the corresponding imputed *_date (first row), but is output as Null currently. Documenting what is being tested will help to identify if the logic from the spec has been met

File: cishouseholds/impute.py (hunk 1 above)
Comment (slamothe99) [reply cishouseholds/impute.py:50]: modified

**Review by ohibo — APPROVED** [review]

**Review by ohibo — APPROVED** [review]

**Review by ohibo — APPROVED** [review]

**Pull request closed without merging**